modules/gapi/test/common/gapi_tests_common.hpp

```diff
@@ -542,6 +542,7 @@ struct TestWithParamsSpecific : public TestWithParamsBase<ParamsSpecific<Specifi
  * @param ...       list of names of user-defined parameters. if there are no parameters, the list
  *                  must be empty.
  */
+ //TODO: Consider to remove `Number` and use `std::tuple_size<decltype(std::make_tuple(__VA_ARGS__))>::value`
```

Comment:
@andrey-golubev any thoughts on this? :)

Reply:
oh, looking at it in a silo, that would actually be nice.  `std::make_tuple` is constexpr since C++14 (is OpenCV on it already?), so in C++11 that won't help.
there's [`BOOST_PP_VARIADIC_SIZE`|https://www.boost.org/doc/libs/1_56_0/boost/preprocessor/variadic/size.hpp] by the way (which is surprisingly interesting), that could work with __VA_ARGS__ as well, but I don't remember this code anymore probably.

Reply:
it depends on whether __VA_ARGS__ are variables or types (which is also a problem with std::make_tuple() I think). you'd need a type function in the latter case.
also note that the logic is to "##" the number to a macro, so that it becomes a new macro name e.g.:
Number = 0: use DEFINE_SPECIFIC_PARAMS_0
Number = 5: use DEFINE_SPECIFIC_PARAMS_5
if I remember this correctly.
so I don't think you'd be able to pull that off with a (possibly-a-type) function to be honest, since the compiler is invoked after the preprocessor.

anyhow, at least you'd need to wrap existing macro into one more to get the literal before using it, which is of course might also be problematic :)

Reply:
in theory, the whole thing could be rewritten to only use templates (making macro a no-op), not sure why I didn't go this way. there was something about declaring member variables which couldn't be done with a simple template magic (maybe with a complicated magic, but it seemed too much at the time probably).

Reply:
>it depends on whether __VA_ARGS__ are variables or types

I think I understand your point: yes - my home-grown function implementation is for variables only, to cover the types we need another solution. But implementation of generic solution (both for types & variables) requires a lot of effort in decomposition sub-problems for types and variables in single overwhelmed entity. 

but for this specific case we receives variables only (for checking fixture correctness), and looks like it is enough to work over `__VA_ARGS__` as variables here

Sorry, if i understood you in a wrong way

Reply:
as I look at it more and more, seems like  `__VA_ARGS__` are actually names of to-be-later-created-variables.
you can look at what [DEFINE_SPECIFIC_PARAMS](https://github.com/opencv/opencv/blob/0f11b1fc0d329c7cd7bd425f006b6583c1f23489/modules/gapi/test/common/gapi_tests_helpers.hpp) does to get the idea.

anyhow, if you figure out a nice way to simplify this, by all means feel free to update this code.